src/metabase/lib/equality.cljc

```diff
@@ -123,362 +110,188 @@
+    {:keys [generous?]}              :- FindMatchingColumnOptions]
+   (case ref-kind
+     ;; Aggregations are referenced by the UUID of the column being aggregated.
+     :aggregation  (->> columns
+                        (filter #(clojure.core/= (:lib/source %) :source/aggregations))
+                        (filter #(clojure.core/= (:lib/source-uuid %) ref-id))
+                        first)
+     ;; Expressions are referenced by name; fields by ID or name.
+     (:expression
+       :field)     (let [plausible (if (string? ref-id)
+                                     (plausible-matches-for-name ref-id (:join-alias opts) columns)
+                                     (plausible-matches-for-id   ref-id (:join-alias opts) columns generous?))]
+                     (case (count plausible)
+                       0 nil
+                       1 (first plausible)
+                       (disambiguate-matches a-ref plausible)))
+     (throw (ex-info "Unknown type of ref" {:ref a-ref}))))
+
+  ([query stage-number a-ref columns]
+   (find-matching-column query stage-number a-ref columns {}))
+
+  ([query                              :- [:maybe ::lib.schema/query]
+    stage-number                       :- :int
+    [_ref-kind _opts ref-id :as a-ref] :- ::lib.schema.ref/ref
+    columns                            :- [:sequential lib.metadata/ColumnMetadata]
+    opts                               :- FindMatchingColumnOptions]
+   (or (find-matching-column a-ref columns opts)
+       (when (and query (number? ref-id))
+         (when-let [resolved (resolve-field-id query stage-number ref-id)]
+           (find-matching-column (assoc a-ref 2 (or (:lib/desired-column-alias resolved)
+                                                    (:name resolved)))
+                                 (for [col columns]
+                                   (if (:id col)
+                                     (assoc col
+                                            :name                     "____lib.equality/noname"
+                                            :lib/desired-column-alias "____lib.equality/noname")
+                                     col))
+                                 opts))))))
+
+(mu/defn find-matching-ref :- [:maybe ::lib.schema.ref/ref]
+  "Given `column` and a list of `refs`, finds the ref that best matches this column.
+
+  Throws if there are multiple, ambiguous matches.
+
+  Returns the matching ref, or nil if no plausible matches are found."
+  [column       :- lib.metadata/ColumnMetadata
+   refs         :- [:sequential ::lib.schema.ref/ref]]
+  (let [ref-tails (group-by last refs)
+        matches   (or (some->> column :lib/source-uuid (get ref-tails) not-empty)
+                      (not-empty (get ref-tails (:id column)))
+                      (not-empty (get ref-tails (:lib/desired-column-alias column)))
+                      (get ref-tails (:name column))
+                      [])]
```

Comment:
It probably could be, but I adopted a policy here of implementing the minimum disambiguation necessary to work in all cases. It turns out the "ref in refs" case is very simple and this never came up. It could almost work with naive `=` matching.